source/Optimizers/nelder_mead.d

```diff
@@ -0,0 +1,519 @@
+                contraction_vector *= _contraction_coef;
+                contraction_vector += centroid_vector;
+            }
+
+            contraction_value = _func(contraction_vector);
+            if (contraction_value < max1_value) {
+                // Readjust centroid vector.
+                temporary_centroid_vector += contraction_vector;
+
+                // Replace last point in simplex.
+                simplex[max1_index].v[] = contraction_vector.v[];
+                simplex_values[max1_index] = contraction_value;
+
+                min_value = simplex_values.minElement;
+                min_index = simplex_values.minIndex;
+                simplex_values.topNCopy!"a > b"(tmp_max_value, Yes.sortOutput);
+                simplex_values.topNIndex!"a > b"(index_max, Yes.sortOutput);
+                max1_value = tmp_max_value[0];
+                max1_index = index_max[0];
+                max2_value = tmp_max_value[1];
+                max2_index = index_max[1];
+                free_pass = true;
+            }
+        }
+
+        /// Shrink
+        if (!free_pass) {
+            // We need to recompute the centroid
+            temporary_centroid_vector.v[] = simplex[min_index].v[];
+            foreach(i; 0 .. num_points) {
+                if (i != min_index) {
+                    simplex[i] -= simplex[min_index];
+                    simplex[i] *= _shrink_coef;
+                    simplex[i] += simplex[min_index];
+                    simplex_values[i] = _func(simplex[i]);
+                    temporary_centroid_vector += simplex[i];
+                }
+            }
+
+            min_value = simplex_values.minElement;
+            min_index = simplex_values.minIndex;
+            simplex_values.topNCopy!"a > b"(tmp_max_value, Yes.sortOutput);
+            simplex_values.topNIndex!"a > b"(index_max, Yes.sortOutput);
+            max1_value = tmp_max_value[0];
+            max1_index = index_max[0];
+            max2_value = tmp_max_value[1];
+            max2_index = index_max[1];
+        }
+
+        // Stopping criterion 1:
+        // Stop when the centroid and the first vector are close
+
+        // Remove worst point from centroid
+        temporary_centroid_vector -= simplex[max1_index];
+
+        temporary_vector.v[] = temporary_centroid_vector.v[];
+        temporary_vector.v[] /= length;
+        temporary_vector -= simplex[min_index];
+        stop_criterion = (temporary_vector.norm!"L2" <= 0.0001);
+
+        ind++;
+    }
+
+    /// Return
+    _v.v[] = simplex[min_index].v[];
+    return min_value;
+}
+unittest {
+    write("Unittest: nelder_mead ... ");
+
+    void test_action_NM(Vector!float expected_solution,
+                        string str_action,
+                        Vector!float point_training=null)
+    {
+        if (!point_training)
+            point_training = expected_solution;
+        size_t len = 2;
+        size_t num_points = 1;

```

Comment:
why is this a constant inside the function?

Reply:
Good call, those are vestigial parameters we need to remove, all the tests are constructed around the fact that we have only one point in dimension 2